src/ast.rs

```diff
@@ -68,39 +65,202 @@ impl<'a> Program<'a> {
-        let mut data = Vec::new();
-        // Create index of all blocks and move out all data block into `data`
+    // TODO: Refactor to use a block_pointer and remove this function
+    fn index_last_block(&mut self) -> Result<(), Error> {
+        let line_number = self.blocks.last().unwrap().line_number();
+        let index = self.blocks.len() - 1;
+        if self.block_index.insert(line_number, index).is_some() {
+            Err(Error::DuplicateLineNumber { line_number })
+        } else {
+            Ok(())
+        }
+    }
+
+    fn build(
+        &mut self,
+        blocks: Vec<Block<'a>>,
+        mut internal_line_numer: u16,
+    ) -> Result<u16, Error> {
+        // 1. Create index of all blocks.
+        // 2. Move out all data block into `data`
         // FIXME: We could increase performance by removing data blocks from all blocks,
         // however we still need to keep index of them, since a GOTO statement could jump
-        // on a data block. For that, we most likely need to use something like skip lists.
-        for (index, block) in blocks.iter_mut().enumerate() {
+        // on a data block. Instead, we replace them with REM.
+        // 3. Flatten FOR blocks and generate equivalent code without FOR.
+        for block in blocks {
             match block {
                 Block::Line {
                     line_number,
-                    ref mut statement,
+                    statement_source,
+                    statement,
                     ..
                 } => {
                     match statement {
-                        Statement::Data(ref mut statement_data) => {
-                            data.append(statement_data);
+                        Statement::Data(mut statement_data) => {
+                            self.data.append(&mut statement_data);
+                            self.blocks.push(Block::Line {
+                                line_number,
+                                statement_source,
+                                statement: Statement::Rem,
+                            })
                         }
-                        _ => (),
+                        other_statement => self.blocks.push(Block::Line {
+                            line_number,
+                            statement_source,
+                            statement: other_statement,
+                        }),
                     };
-                    if block_index.insert(*line_number, index).is_some() {
-                        return Err(Error::DuplicateLineNumber {
-                            line_number: *line_number,
+                    self.index_last_block()?;
+                }
+                // Flatten FOR...NEXT by replacing FOR statement with LET preamble and
+                // IF statement, NEXT statement with GOSUB.
+                Block::For {
+                    for_line,
+                    blocks,
+                    next_line,
+                } => {
```

Comment:
What do you think about factoring out the statements? I think this would make reading it easier.

Reply:
I will create an issue for that. You are right, this needs to be cleaned up.